packages/mysql/lib/mysql_p.d.ts

```diff
@@ -72,4 +74,69 @@ declare namespace captureMySQL {
     ? PatchedCreatePoolClusterFunction
     : T[K];
   };
+
+  type MySQL2QueryCallbackResultArg =
+  | MySQL2.RowDataPacket[][]
+  | MySQL2.RowDataPacket[]
+  | MySQL2.OkPacket
+  | MySQL2.OkPacket[]
+  | MySQL2.ResultSetHeader;
+
+  type MySQL2QueryCallback = <T extends MySQL2QueryCallbackResultArg>(
+      err: MySQL2.QueryError | null,
+      result: T,
+      fields: MySQL2.FieldPacket[],
+    ) => any;
+
+  interface PatchedMySQL2PoolGetConnectionFunction {
+    (): Promise<PatchedConnection>;
+  }
+
+  type PatchedMySQL2Pool<T = MySQL2.Pool> = {
+    [K in keyof T]: K extends 'query'
+    ? PatchedQueryFunction
+    : K extends 'getConnection'
+    ? PatchedMySQL2PoolGetConnectionFunction
+    : T[K];
+  };
+
+  interface PatchedMySQL2PoolClusterOfFunction {
+    (pattern: string, selector?: string): PatchedMySQL2Pool;
+  }
+
+  interface PatchedMySQL2PoolClusterGetConnectionFunction {
+    (): Promise<PatchedConnection>;
+    (group: string): Promise<PatchedConnection>;
+    (group: string, selector: string): Promise<PatchedConnection>;
+  }
+
+  type PatchedMySQL2PoolCluster<T = MySQL2.PoolCluster> = {
+    [K in keyof T]: K extends 'of'
+    ? PatchedMySQL2PoolClusterOfFunction
+    : K extends 'getConnection'
+    ? PatchedMySQL2PoolClusterGetConnectionFunction
+    : T[K]
+  };
+
+  interface PatchedMySQL2CreateConnectionFunction {
+    (config: MySQL2.ConnectionOptions | string): Promise<PatchedConnection>;
```

Comment:
By default, we should also support `mysql2.createConnection` that returns a `Connection`. Our instrumentation does support a return value of a `Promise<Connection>`, but that requires supporting TS definitions of patched `mysql2/promise` rather than `mysql2`, which only the latter is supported by this PR.

Reply:
I've de-scoped these changes to support return of `Connection` only, rather than extending it to support `Promise<Connection>`: [8d4597f](https://github.com/aws/aws-xray-sdk-node/pull/676/commits/8d4597ff4da50ac7acb499405e95112021ffd0a9)

Reply:
Thanks @jj22ee for calling out the distinction between `mysql2` and `mysql2/promise`, which I didn't quite appreciate on the outset.

I'd be happy to explore an extension of X-Ray's type definitions for the latter in future, but I'll keep this PR specific to `mysql2` for the time being.